source-boilerplate/boilerplate.go

```diff
@@ -278,3 +281,82 @@ func (a *pullAdapter) SetHeader(metadata.MD) error {
 func (a *pullAdapter) SetTrailer(metadata.MD) {
 	panic("SetTrailer is not supported")
 }
+
+// PullOutput is a convenience wrapper around a pc.Driver_PullServer which makes it simpler
+// to emit documents and state checkpoints.
+type PullOutput struct {
+	sync.Mutex
+
+	Stream pc.Driver_PullServer
+}
+
+// Ready sends a PullResponse_Opened message to indicate that the capture has started.
+func (out *PullOutput) Ready() error {
+	log.Debug("sending PullResponse.Opened")
+	out.Lock()
+	defer out.Unlock()
+	if err := out.Stream.Send(&pc.PullResponse{Opened: &pc.PullResponse_Opened{}}); err != nil {
+		return fmt.Errorf("error sending PullResponse.Opened: %w", err)
+	}
+	return nil
+}
+
+// Documents emits one or more documents to the specified binding index.
+func (out *PullOutput) Documents(binding uint32, docs ...json.RawMessage) error {
+	log.WithField("count", len(docs)).Trace("emitting documents")
+
+	// Concatenate multiple documents into the arena, with appropriate indices
+	var arena []byte
+	var slices []pf.Slice
+	for _, doc := range docs {
+		var begin = uint32(len(arena))
+		arena = append(arena, []byte(doc)...)
+		slices = append(slices, pf.Slice{Begin: begin, End: uint32(len(arena))})
+	}
+
+	// TODO(wgd): The logic of this function needs to be rewritten to support
+	//   splitting into multiple 'Documents' RPCs in order to keep the size of
+	//   any individual message bounded.
+	if len(arena) > 2*1024*1024 {
```

Comment:
Is this a new limitation? I'm not sure if it was part of airbyte-to-flow as well, or if it is something new that might happen.

Reply:
I'm not sure if it's an airbyte-to-flow problem or not, it depends on how the relevant translation is implemented there. It's possible to just implement the output logic in a dumber way where multiple documents never get combined into a single Documents RPC...and now that I've typed that it seems obvious that I should just remove the "coalesce multiple documents" logic for now and just implement this function in the less-theoretically-efficient [1] way.

[1] Neither `source-firestore` nor the `sqlcapture` package ever emit multiple documents via a single call so the efficiency argument is irrelevant in a practical sense at the moment.